File: SpearoGo/Services/WeatherService.swift
import Foundation
import CoreLocation

struct WeatherService {
    // Open-Meteo weather API — free, no key, global coverage
    // Docs: https://open-meteo.com/en/docs

    private let session: URLSession = {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 15
        config.timeoutIntervalForResource = 30
        return URLSession(configuration: config)
    }()

    func fetch(coordinate: CLLocationCoordinate2D) async throws -> WeatherData {
        let url = try buildURL(coordinate: coordinate)
        let (data, response) = try await session.data(from: url)

        guard let http = response as? HTTPURLResponse, http.statusCode == 200 else {
            throw ServiceError.badResponse
        }

        let decoded = try JSONDecoder().decode(OpenMeteoWeatherResponse.self, from: data)
        guard let current = decoded.current else { throw ServiceError.missingData }

        // Wind is the one reading this app cannot do without, and a missing
        // wind speed must not become a flat calm — a 0 kn default reads as
        // ideal conditions. Everything else is carried through as optional and
        // rendered as "—" rather than filled in.
        guard let windMS = current.wind_speed_10m,
              let windDir = current.wind_direction_10m else {
            throw ServiceError.missingData
        }

        // Convert m/s → knots
        let windKnots  = windMS * 1.94384
        let gustsKnots = current.wind_gusts_10m.map { $0 * 1.94384 }

        let daily = decoded.daily

        return WeatherData(
            windSpeed:     windKnots,
            windDirection: windDir,
            windGusts:     gustsKnots ?? windKnots,
            visibility:    current.visibility.map { $0 / 1000 },
            cloudCover:    current.cloud_cover,
            airTemp:       current.temperature_2m,
            tempMax:       daily?.temperature_2m_max?.first ?? nil,
            tempMin:       daily?.temperature_2m_min?.first ?? nil,
            precipitationChance: daily?.precipitation_probability_max?.first.flatMap { $0 },
            weatherCode:   daily?.weather_code?.first ?? current.weather_code,
            fetchedAt:     Date()
        )
    }

    private func buildURL(coordinate: CLLocationCoordinate2D) throws -> URL {
        guard var components = URLComponents(string: Constants.API.weatherBase) else {
            throw ServiceError.invalidURL
        }
        components.queryItems = [
            .init(name: "latitude",         value: String(coordinate.latitude)),
            .init(name: "longitude",        value: String(coordinate.longitude)),
            .init(name: "current",          value: "wind_speed_10m,wind_direction_10m,wind_gusts_10m,cloud_cover,visibility,temperature_2m,weather_code"),
            .init(name: "daily",            value: "temperature_2m_max,temperature_2m_min,precipitation_probability_max,weather_code"),
            .init(name: "wind_speed_unit",  value: "ms"),
            .init(name: "timezone",         value: "auto"),
            .init(name: "forecast_days",    value: "1")
        ]
        guard let url = components.url else {
            throw ServiceError.invalidURL
        }
        return url
    }
}

// MARK: - Response shapes

private struct OpenMeteoWeatherResponse: Decodable {
    let current: CurrentWeather?
    let daily: DailyWeather?
}

private struct CurrentWeather: Decodable {
    let wind_speed_10m:      Double?
    let wind_direction_10m:  Double?
    let wind_gusts_10m:      Double?
    let cloud_cover:         Int?
    let visibility:          Double?
    let temperature_2m:      Double?
    let weather_code:        Int?
}

private struct DailyWeather: Decodable {
    let temperature_2m_max:           [Double]?
    let temperature_2m_min:           [Double]?
    let precipitation_probability_max: [Int?]?
    let weather_code:                 [Int]?
}

enum ServiceError: Error {
    case badResponse
    case missingData
    /// The coordinate has no sea: the marine API answered, with nothing in it.
    case noMarineCoverage
    case invalidCoordinate
    case invalidURL
}

File: wear/app/src/main/java/com/spearotracker/spearogo/services/WeatherService.kt
package com.spearotracker.spearogo.services

import com.spearotracker.spearogo.models.WeatherData
import com.spearotracker.spearogo.utils.Constants
import okhttp3.OkHttpClient
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import retrofit2.http.GET
import retrofit2.http.Query
import java.util.concurrent.TimeUnit
import javax.inject.Inject
import javax.inject.Singleton

private interface OpenMeteoWeatherApi {
    @GET("v1/forecast")
    suspend fun forecast(
        @Query("latitude") latitude: Double,
        @Query("longitude") longitude: Double,
        @Query("current") current: String = "wind_speed_10m,wind_direction_10m,wind_gusts_10m,cloud_cover,visibility,temperature_2m,weather_code",
        @Query("daily") daily: String = "temperature_2m_max,temperature_2m_min,precipitation_probability_max,weather_code",
        @Query("wind_speed_unit") windSpeedUnit: String = "ms",
        @Query("timezone") timezone: String = "auto",
        @Query("forecast_days") forecastDays: Int = 1
    ): OpenMeteoWeatherResponse
}

private data class OpenMeteoWeatherResponse(
    val current: CurrentWeather?,
    val daily: DailyWeather?
)

private data class CurrentWeather(
    val wind_speed_10m: Double?,
    val wind_direction_10m: Double?,
    val wind_gusts_10m: Double?,
    val cloud_cover: Int?,
    val visibility: Double?,
    val temperature_2m: Double?,
    val weather_code: Int?
)

private data class DailyWeather(
    val temperature_2m_max: List<Double?>?,
    val temperature_2m_min: List<Double?>?,
    val precipitation_probability_max: List<Int?>?,
    val weather_code: List<Int?>?
)

@Singleton
class WeatherService @Inject constructor() {

    private val client = OkHttpClient.Builder()
        .connectTimeout(10, TimeUnit.SECONDS)
        .readTimeout(15, TimeUnit.SECONDS)
        .build()

    private val api: OpenMeteoWeatherApi = Retrofit.Builder()
        .baseUrl(Constants.Api.WEATHER_BASE)
        .client(client)
        .addConverterFactory(GsonConverterFactory.create())
        .build()
        .create(OpenMeteoWeatherApi::class.java)

    suspend fun fetch(latitude: Double, longitude: Double): WeatherData {
        val response = api.forecast(latitude, longitude)
        val current = response.current ?: throw ServiceException("Missing weather data")

        // Wind is the one reading this app cannot do without, and a missing
        // wind speed must not become a flat calm - a 0 kn default reads as
        // ideal conditions. Everything else is carried through as nullable and
        // rendered as "—" rather than filled in.
        val windMS = current.wind_speed_10m
        val windDir = current.wind_direction_10m
        if (windMS == null || windDir == null) throw ServiceException("Missing wind data")

        // Convert m/s to knots
        val windKnots = windMS * 1.94384
        val gustsKnots = current.wind_gusts_10m?.times(1.94384)

        val daily = response.daily

        return WeatherData(
            windSpeed = windKnots,
            windDirection = windDir,
            windGusts = gustsKnots ?: windKnots,
            visibility = current.visibility?.div(1000.0),
            cloudCover = current.cloud_cover,
            airTemp = current.temperature_2m,
            tempMax = daily?.temperature_2m_max?.firstOrNull(),
            tempMin = daily?.temperature_2m_min?.firstOrNull(),
            precipitationChance = daily?.precipitation_probability_max?.firstOrNull(),
            weatherCode = daily?.weather_code?.firstOrNull() ?: current.weather_code
        )
    }
}
